### ISSUE TITLE
possible to use ReverseSearch from R5 in travel_time_matrix?

### ISSUE THREAD

**Issue by LeLunZ** [ipeaGIT/r5r#291]
Hello :)

i browsed through the actual R5 Repository and found out that there is a ReverseSearch attribute in the ProfileRequest. 

Is it possible to use that attribute with the travel time matrix function? 

For calculating the travel time as arrive by matrix and not departure by?

**Comment by LeLunZ** [comment]
After talking with a colleague of mine, I have tried a thing.

I found this guide: https://cran.r-project.org/web/packages/r5r/vignettes/calculating_isochrones.html
and I just reversed origin and destination points, which also gave me the result I wanted (I think)

The thing is instead of doing just one iterations, I know have N iterations :) which isn't that nice 
Would it be somehow possible to reverse edge direction? and do it with one iteration?

For more context:
I would like to be able to get a isochrone for up to a 200 kilometer radius.

I actually first found the R5 [analysis endpoint](https://github.com/conveyal/r5/blob/43068e168138bc5636d895861209c1c0049e0f1a/src/main/java/com/conveyal/analysis/controllers/BrokerController.java#L126) which gives me nice isochrones if I use the Header `Accept: image/tiff`

So, R5 can't do the reverse thing. That's when I found the R5R package. 
Is the travel_time_matrix function even the correct function if I want to do this for such a big area?


I would appreciate any help :)

**Comment by rafapereirabr** [comment]
Hi @LeLunZ . I'm sorry for my late reply.  I'm not entirely certain yet, but I believe it might be possible to implement reverseSearch in `r5r`, so that users can input an `arrial_time` instead of a `departure_time`.  We will be looking into this in the next few months. Please keep in mind this is not top priority at the moment, it might take some time before we are able to investigate this further

**Comment by Torskelgen** [comment]
Hi @rafapereirabr! Any worthwhile status update on using `arrival_time `instead of `departure_time`? And keep up with the amazing work you do with this package!

**Comment by rafapereirabr** [comment]
Hi @Torskelgen . Thanks for asking. We are currently prioritizing other issues, so it might take a few months before we can tackle this one. Sorry we have our hands tight at the moment. If you have any experience with Jave and feel like you could contribute with this, we would be glad to accept contributions.

**Comment by Torskelgen** [comment]
I'm afraid I don't have any experience with Java, otherwise I would have loved to contribute. And I do understand that you prioritize other issues.

**Comment by rafapereirabr** [comment]
If there are any potential contributors who know Java and who might want to have a look at this, we would appreciate PR from collaborators

**Comment by Henrqe** [comment]
Hello,

Me and my friend are trying to make a isochrone for multiple points using r5r. However, we are not confident with the step-by-step since we are not very skilled in programming and to the best of our knowledge r5r only generates the isochrone as from one point, right?! So, we adapted the alternative example provided in the [**4. topic isochrone vignette**](https://cran.r-project.org/web/packages/r5r/vignettes/isochrones.html) (as it follows) and we would like to know if it's reasonable way to use r5r to calculate a isochrone for multiple points:

**The tip provided by @LeLunZ of switch the origin to be the point of interest and the destinations as the centroids seems to make a better visualization as of the inverse way.**

\#
options(java.parameters = "-Xmx4G")

library(r5r)
library(sf)
library(data.table)
library(ggplot2)
library(interp)
library(fs)
library(tidyverse)

\# set data path to directory containing your own data if not running this example
data_path <- system.file("extdata/poa", package = "r5r")

\#
r5r_core <- setup_r5(data_path)
dir_tree(data_path)

\# read all points in the city
points <- fread(file.path(data_path, "poa_hexgrid.csv"))

\# read all points of interest in the city
pois <- fread(file.path(data_path, "poa_points_of_interest.csv"))

\# select multiple points of interest that are distant of each other
distantpois <- c("bus_central_station", "pucrs", "praia_de_belas_shopping_center" )
poisfiltered <- pois %>% filter(id %in% distantpois)

\# routing inputs
mode <- c("CAR")
max_trip_duration <- 100 # in minutes
departure_datetime <- as.POSIXct("13-05-2019 10:00:00",
                                 format = "%d-%m-%Y %H:%M:%S")

\# calculate travel time matrix
ttm <- travel_time_matrix(r5r_core,
                          origins = poisfiltered,
                          destinations = points,
                          mode = mode,
                          departure_datetime = departure_datetime,
                          #max_walk_time = max_walk_time,
                          max_trip_duration = max_trip_duration,
                          #time_window = time_window,
                          progress = FALSE)

head(ttm)

\# add coordinates of destinations to travel time matrix
ttm[points, on=c('to_id' ='id'), `:=`(lon = i.lon, lat = i.lat)]

\# now we have a travel time matrix from each of the 3 points of interest selected
\# identifying duplicate coordinates of centroids since we run the ttm for more than 1 poi
duplicated_coords <- ttm[, .N, by = .(lon, lat)][N > 1]
print(duplicated_coords)

\# filtering the least travel time value from travel_time_p50 to leave just the value to the nearest POI 
ttm_filtered <- ttm[, .SD[which.min(travel_time_p50)], by = .(lon, lat)]
  \# just checking if the duplicates were already removed, for the case of other city
  duplicated_coords <- ttm_filtered[, .N, by = .(lon, lat)][N > 1]

\# Exibir o resultado
print(duplicated_coords) #seems ok! no duplicates
print(ttm_filtered) #ttm from 3 POIS 

\# extract OSM network
street_net <- street_network_to_sf(r5r_core)
edges <- subset(street_net$edges, street_class %like% 'PRIMARY|SECONDARY')

\# interpolate estimates to get spatially smooth result
travel_times.interp <- with(na.omit(ttm_filtered), interp(lon, lat, travel_time_p50)) |>
  with(cbind(travel_time=as.vector(z),  # Column-major order
             x=rep(x, times=length(y)),
             y=rep(y, each=length(x)))) |>
  as.data.frame() |> na.omit()

\# find isochrone's bounding box to crop the map below
bb_x <- c(min(travel_times.interp$x), max(travel_times.interp$x))
bb_y <- c(min(travel_times.interp$y), max(travel_times.interp$y))

\# plot
ggplot(travel_times.interp) +
  geom_sf(data = edges, color = "gray55", size=0.01, alpha = 0.7) +
  geom_contour_filled(aes(x=x, y=y, z=travel_time), alpha=.7) +
  geom_point(aes(x=lon, y=lat, color='Points of\ninterest'),
             data=poisfiltered) +
  scale_fill_viridis_d(direction = -1, option = 'A') +
  scale_color_manual(values=c('Points of\ninterest'='black')) +
  scale_x_continuous(expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  coord_sf(xlim = bb_x, ylim = bb_y) +
  labs(fill = "Travel time\n(in minutes)", color='') +
  theme_minimal() +
  theme(axis.title = element_blank())

r5r::stop_r5(r5r_core)
rJava::.jgc(R.gc = TRUE)

Second, this alternative seems to generate only a regular grid with long/lat with its respective interpolated values. This points seems to be conected through the **ggplot::geom_countour_filled**. However, it seems to only generate a static image (jpeg, png, tiff etc.). We would like to know if there is a way to export the generated polygons (like vectors) in .gpkg format to use in QGIS the same way it is generated automatically with the isochrone() function (we think). It can be also a way to export the image in raster format with geocoded information, in tiff format for example, that QGIS could be able to recognize. The attached image is generated with the script above.

We are trying to make it work in r5r due its replicability potential. Sorry for the long text!

![Image](https://github.com/user-attachments/assets/76c67867-d2b6-46e5-bd24-7c936a211835)

**Comment by rafapereirabr** [comment]
Hi @Henrqe , this issue 291 is about using _reverseSearch_ in {r5r}. Please open a separate issue for your question.

**Comment by BardyBard** [comment]
Hi @rafapereirabr, I took a look at the R5 code and how it ties together with R5R. 

The reverse search property mentioned by the user above is a reverse search not an arrive by, it is used for the last step of the routing, finding the closest transit stops from the destination. 

As for an arrive by: R5R currently implements TimeTravelMarix with R5 FastRaptorWorkers? Although this process has an arrive_by property it seems to have never been implemented by Conveyal. According to their documentation, they have new workers called otp2, these have an arriveByTransit() method that seems to be implemented. I can try to investigate this issue and see how difficult it would be to migrate the TimeTravelMatrix to these newer workers.

In theory this would be a necessary transition for R5R at some point as this otp2 process is what I assume conveyal is transitioning to. Since this would require a re implementation of TimeTravelMatrix it would also erase the code redundancy in  [#386](https://github.com/ipeaGIT/r5r/issues/386). 

This sounds like a very big task, I'll look into it in more detail later this week.

**Comment by rafapereirabr** [comment]
Hi @ansoncfit and @mattwigway, we've been investigating the possibility of including in {r5r} and `arrival_time` argument. So instead of choose the departure time of the trip, users could select the arrival time. I believe this is implemented in OTP, but I'm not whether something similar in present in R5. Would you have any insights here? @BardyBard is helping us with a few issues in {r5r} and any insights here would be much appreciated.

**Comment by mattwigway** [comment]
We've discussed this previously. We initially planned to implement something similar to the otp approach in r5, where you could specify an arrival time and it would search backwards from the destination. This was really difficult to maintain in OTP because there was a huge amount of conditional logic for whether a search was happening forwards or backwards. Our first thought to solve this was to reverse the network instead of the algorithm (conveyal/r5#253. We later realized that even this isn't necessary, and you can get the same results just by carefully choosing departure times and filteringn results from an efficient range-RAPTOR implementation (see the penultimate comment on conveyal/r5#785, and the final paragraph of [this working paper] (https://projects.indicatrix.org/range-raptor-transfer-compression/) to see how this would work. This latter approach is what I use in [TransitRouter.jl](https://GitHub.com/mattwigway/TransitRouter.jl) and it works well. The quick and dirty way to do this would be to just run a search for significantly longer window than desired and filter, which could be implemented without any Java changes, I think; being smarter about stopping rules would require some changes to r5 itself, but that would only be necessary to improve performance.

**Comment by BardyBard** [comment]
@rafapereirabr implementing a proper reverse search for detailed itineraries would require massive duplications of R5 code.

I will now try to look at whether it's possible with `time_travel_matrix` without a large duplication.

**Comment by rafapereirabr** [comment]
This issue has now been solved with https://github.com/ipeaGIT/r5r/pull/446. The `arrival_travel_time_matrix()` function is currently implemented in the dev version and will be included in the next release on CRAN.
